File: PyConfig/openwns/Buffer.py
###############################################################################
# This file is part of openWNS (open Wireless Network Simulator)
# _____________________________________________________________________________
#
# Copyright (C) 2004-2007
# Chair of Communication Networks (ComNets)
# Kopernikusstr. 5, D-52074 Aachen, Germany
# phone: ++49-241-80-27910,
# fax: ++49-241-80-22242
# email: info@openwns.org
# www: http://www.openwns.org
# _____________________________________________________________________________
#
# openWNS is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License version 2 as published by the
# Free Software Foundation;
#
# openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################

from openwns.pyconfig import attrsetter
import openwns.logger
import openwns.FUN

class Buffer(openwns.FUN.FunctionalUnit):
    size = 1
    sizeUnit = 'PDU'
    lossRatioProbeName = 'NoNameSpecified'
    droppedPDUWindowDuration = 0.20 # seconds
    probeDroppedPDUInterval  = 0.10 # seconds
    sizeProbeName = 'NoNameSpecified'
    logger = None
    probingEnabled = True
    localIDs = None

    def setLocalIDs(self, localIDs):
        self.localIDs = localIDs

    def addLocalIDs(self, localIDs):
        self.localIDs.update(localIDs)

    def __init__(self, parentLogger = None, **kw):
        super(Buffer, self).__init__()
        self.localIDs = {}
        self.logger = openwns.logger.Logger("WNS", "Buffer", True, parentLogger)
        attrsetter(self, kw)

class Bounded(Buffer):
    __plugin__ = 'wns.buffer.Bounded'
    name = "Bounded"
    def __init__(self, **kw):
        super(Bounded, self).__init__(**kw)

class Dropping(Buffer):
    __plugin__ = 'wns.buffer.Dropping'
    name = "Dropping"

    drop = 'Tail'
    def __init__(self, **kw):
        super(Dropping, self).__init__(**kw)

File: src/ldk/buffer/Buffer.cpp
/*******************************************************************************
 * This file is part of openWNS (open Wireless Network Simulator)
 * _____________________________________________________________________________
 *
 * Copyright (C) 2004-2007
 * Chair of Communication Networks (ComNets)
 * Kopernikusstr. 5, D-52074 Aachen, Germany
 * phone: ++49-241-80-27910,
 * fax: ++49-241-80-22242
 * email: info@openwns.org
 * www: http://www.openwns.org
 * _____________________________________________________________________________
 *
 * openWNS is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 2 as published by the
 * Free Software Foundation;
 *
 * openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

#include <WNS/ldk/fun/FUN.hpp>
#include <WNS/ldk/Layer.hpp>
#include <WNS/ldk/buffer/Buffer.hpp>
#include <WNS/probe/bus/utils.hpp>

using namespace wns::ldk::buffer;
using namespace wns::ldk::buffer::sizecalculators;

// Size calculation stategies
unsigned long int
PerPDU::operator()(const CompoundPtr& /* pdu */) const
{
	return 1;
}
STATIC_FACTORY_REGISTER(PerPDU, SizeCalculator, "PDU");

unsigned long int
PerBit::operator()(const CompoundPtr& compound) const
{
	return compound->getLengthInBits();
}
STATIC_FACTORY_REGISTER(PerBit, SizeCalculator, "Bit");



Buffer::Buffer(fun::FUN* _fun, const pyconfig::View& _config) :
	events::PeriodicTimeout(),
	totalPDUs(0),
	droppedPDUs(0),
	droppedPDUWindow(_config.get<double>("droppedPDUWindowDuration")),
	probeDroppedPDUInterval(_config.get<double>("probeDroppedPDUInterval")),
	logger(_config.get("logger")),
	probingEnabled(_config.get<bool>("probingEnabled"))
{
	// read the localIDs from the config
	wns::probe::bus::ContextProviderCollection localContext(_fun->getLayer()->getContextProviderCollection());
	for (int ii = 0; ii<_config.len("localIDs.keys()"); ++ii)
	{
		std::string key = _config.get<std::string>("localIDs.keys()",ii);
		unsigned long int value  = _config.get<unsigned long int>("localIDs.values()",ii);
		localContext.addProvider( wns::probe::bus::contextprovider::Constant(key, value) );
	}

	std::string sizeProbeName = _config.get<std::string>("sizeProbeName");
	std::string lossProbeName = _config.get<std::string>("lossRatioProbeName");
	lossRatioProbeBus = wns::probe::bus::ContextCollectorPtr(new wns::probe::bus::ContextCollector(localContext, lossProbeName));
	sizeProbeBus      = wns::probe::bus::ContextCollectorPtr(new wns::probe::bus::ContextCollector(localContext, sizeProbeName));
}

Buffer::~Buffer()
{
}

void
Buffer::onFunCreated()
{
	// is not called. Whyever!?
}

void
Buffer::increaseTotalPDUs()
{
	// start on first seen compound
	if (probingEnabled)
	{
		// start on first seen compound
		if ((this->hasPeriodicTimeoutSet() == false) && (probeDroppedPDUInterval>0.0))
		{
			MESSAGE_SINGLE(NORMAL, logger,"Buffer::increaseTotalPDUs: startPeriodicTimeout(every "<<probeDroppedPDUInterval<<"s)");
			this->startPeriodicTimeout(this->probeDroppedPDUInterval);
		}
	}
	++this->totalPDUs;
}


void
Buffer::increaseDroppedPDUs(int size)
{
	++this->droppedPDUs;
	this->droppedPDUWindow.put(size);
}

void
Buffer::periodically()
{
	if (this->lossRatioProbeBus)
	{
		this->lossRatioProbeBus->put(this->droppedPDUWindow.getPerSecond());
	}
}

void
Buffer::probe()
{
	if(this->sizeProbeBus)
	{
		this->sizeProbeBus->put(((double) getSize()) / getMaxSize());
	}
} // probe



